In [13]:
from sklearn.datasets import load_digits, load_wine, load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score

### 데이터 준비

In [14]:
# 데이터 이해하기
dataset = load_digits()
digits = dataset.data
label = dataset.target
print(digits.shape)
print(dataset.feature_names)
print(dataset.target_names)

import pandas as pd
df = pd.DataFrame(data=digits, columns=dataset.feature_names)
df.head()

(1797, 64)
['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']
[0 1 2 3 4 5 6 7 8 9]


pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [15]:
print(dataset.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

[데이터 이해]    
손으로 쓴 숫자를 시각적으로 인식하는 문제. UCI에서 만든 수기 숫자 데이터셋 중에서 테스트셋이다.   
레코드의 수는 1797개이다. 특성값은 64개로, 8x8 이미지의 각 픽셀을 나타낸다. 한 픽셀의 값은 0과 16 사이의 정수이다. 원본 이미지는 32x32 사이즈인데 이것을 8x8 사이즈로 축소하였다. 32x32 이미지를 4x4 사각형들로 겹치지 않게 나눈 다음, 나눠진 사각형 안의 16개 픽셀 중 값이 있는 픽셀의 수를 세어 축소된 한 픽셀의 값으로 할당하였다.    
레이블은 이미지가 나타내는 숫자로, 0과 9 사이의 정수이다.   


In [16]:
# train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits, label, test_size=0.2, random_state=777)

In [17]:
# 다양한 모델로 학습시켜보기
from sklearn.tree import DecisionTreeClassifier
decision_tree = DecisionTreeClassifier(random_state=2022)
decision_tree.fit(X_train, y_train)
y_pred1 = decision_tree.predict(X_test)
print(accuracy_score(y_test, y_pred1))


0.85


[] 궁금: 이렇게 5개 모델들에서 튜닝할 거리가 무엇이 있을까?

In [18]:
from sklearn.ensemble import RandomForestClassifier
random_forest = RandomForestClassifier(random_state=2022)
random_forest.fit(X_train, y_train)
y_pred2 = random_forest.predict(X_test)
print(accuracy_score(y_test, y_pred2))

0.9638888888888889


In [19]:
from sklearn import svm
svm_model = svm.SVC(random_state=2022)
svm_model.fit(X_train, y_train)
y_pred3 = svm_model.predict(X_test)
print(accuracy_score(y_test, y_pred3))

0.9805555555555555


In [20]:
from sklearn.linear_model import SGDClassifier
sgd = SGDClassifier(random_state=2022)
sgd.fit(X_train, y_train)
y_pred4 = sgd.predict(X_test)
print(accuracy_score(y_test, y_pred4))

0.9222222222222223


In [21]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(random_state=2022, max_iter=5000)
model.fit(X_train, y_train)
pred5 = model.predict(X_test)
print(accuracy_score(y_test, pred5))

0.9472222222222222


digits dataset은 전반적으로 분류를 맞게 매기는게 중요한 문제다. 따라서 전체 추론 중 맞는 추론의 개수를 세는 accuracy로 성능을 평가하는 것이 적합하다. SVM이 98%로 가장 높은 성능을 보였다.

(튜닝을 더 해볼 여지가 있는가?)
(0~16인 데이터 범위를 normalize해주는게 나을까?)